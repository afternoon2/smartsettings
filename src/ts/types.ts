import { Omit } from 'utility-types';

export type ListenerUpdate = {
  targetId: string,
  key: string,
  value: string | boolean | number | DropDownItem[] | DropDownItem,
  listenerType: 'control' | 'section' | 'panel' | 'builtin',
};

export type Listener = (update: ListenerUpdate) => void;

export type RootOptions = {
  name: string,
  invisible?: boolean,
  disabled?: boolean,
  listener?: Listener,
};

export type SectionOptions = RootOptions & {
  collapsed?: boolean;
};

export type SectionState = Omit<SectionOptions, 'listener'>;

export type PanelOptions = SectionOptions & {
  draggable?: boolean,
  top?: number,
  left?: number,
};

type PanelState = Omit<PanelOptions, 'listener'>;

export type TextControlOptions = RootOptions & {
  value?: string,
  autocomplete?: boolean,
  maxLength?: number,
  pattern?: string,
  placeholder?: string,
  readOnly?: boolean,
};

type TextControlState = Omit<TextControlOptions, 'listener'>;

export type TextAreaControlOptions = RootOptions & {
  value?: string,
  autocomplete?: boolean,
  cols?: number,
  maxLength?: number,
  minLength?: number,
  placeholder?: string,
  readOnly?: boolean,
  rows?: number,
};

type TextAreaControlState = Omit<TextAreaControlOptions, 'listener'>;

export type ButtonControlOptions = RootOptions & {
  type?: string,
};

type ButtonControlState = Omit<ButtonControlOptions, 'listener'>;

export type CheckboxControlOptions = RootOptions & {
  checked?: boolean,
  readOnly?: boolean,
};

type CheckboxControlState = Omit<CheckboxControlOptions, 'listener'>;

export type NumberControlOptions = RootOptions & {
  value?: number,
  max?: number,
  min?: number,
  placeholder?: string,
  readOnly?: boolean,
  step?: number,
};

type NumberControlState = Omit<NumberControlOptions, 'listener'>;

export type FileControlOptions = RootOptions & {
  files?: FileList,
  accept?: string,
  disabled?: boolean,
  // multiple?: boolean,
  readOnly?: boolean,
};

type FileControlState = Omit<FileControlOptions, 'listener'>;

export type RangeControlOptions = RootOptions & {
  value?: number,
  max?: number,
  min?: number,
  readOnly?: boolean,
  step?: number,
};

type RangeControlState = Omit<RangeControlOptions, 'listener'>;

export type DropDownItem = {
  id: string,
  text: string,
  value: string,
};

export type DropDownControlOptions = RootOptions & {
  expanded?: boolean,
  selected?: string,
  items?: DropDownItem[],
  // multi?: boolean,
};

type DropDownControlState = Omit<DropDownControlOptions, 'listener'>;

export type ColorControlOptions = RootOptions & {
  value?: string,
  expanded?: boolean,
};

type ColorControlState = Omit<ColorControlOptions, 'listener'>;

export type ControlOptions = ButtonControlOptions
| TextControlOptions
| CheckboxControlOptions
| NumberControlOptions
| FileControlOptions
| TextAreaControlOptions
| RangeControlOptions
| DropDownControlOptions
| ColorControlOptions;

type ControlState = ButtonControlState
| TextControlState
| CheckboxControlState
| NumberControlState
| FileControlState
| TextAreaControlState
| RangeControlState
| DropDownControlState
| ColorControlState;

export type InternalState = (PanelState | ControlState) & {
  [key: string]: string | boolean | number | DropDownItem[],
  readonly id: string,
};

export type InternalStateSetter = (
  target: InternalState,
  key: string,
  value: string | boolean | number,
) => boolean;

export type ParentOptions = PanelOptions | SectionOptions;

export type PanelPosition = {
  top: number,
  left: number,
};

export type ConfigControlNode = Omit<InternalState, 'id'> & {
  displayType: string,
};

export type ConfigSectionOptions = SectionOptions & {
  displayType: string,
};

export type ConfigSectionNode = {
  options: ConfigSectionOptions,
  children: {
    [key: string]: ConfigControlNode,
  },
};

export type SectionConfig = {
  [key: string]: ConfigControlNode
};

export type PanelConfig = {
  [key: string]: ConfigControlNode | ConfigSectionNode
};
