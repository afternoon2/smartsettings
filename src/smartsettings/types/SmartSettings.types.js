// @flow

type PositionObject = {
    top: number,
    left: number
}

type SmartSettingsObject = {
    name: string,
    position: PositionObject
}

export type { PositionObject, SmartSettingsObject }