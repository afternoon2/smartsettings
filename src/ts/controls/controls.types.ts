export type ControlType = 'button'
| 'range'
| 'select'
| 'checkbox'
| 'text'
| 'textarea'
| 'number'
| 'file'
| 'color';

export type TextControlOptions = {
  value: string,
  autocomplete?: boolean,
  autofocus?: boolean,
  disabled?: boolean,
  maxLength?: number,
  name?: string,
  pattern?: string,
  placeholder?: string,
  readOnly?: boolean,
};

export type TextAreaControlOptions = {
  value: string,
  autocomplete?: boolean,
  autofocus?: boolean,
  cols?: number,
  disabled?: boolean,
  maxLength?: number,
  minLength?: number,
  placeholder?: string,
  readOnly?: boolean,
  rows?: number,
  wrap?: boolean,
};

export type CheckboxControlOptions = {
  value: string,
  alt?: string,
  autocomplete?: boolean,
  autofocus?: boolean,
  checked?: boolean,
  disabled?: boolean,
  maxLength?: number,
  pattern?: string,
  placeholder?: string,
  readOnly?: boolean,
};

export type ButtonControlOptions = {
  value: string,
  autofocus?: boolean,
  disabled?: boolean,
  type?: string,
};

export type NumberControlOptions = {
  value: number,
  autofocus?: boolean,
  disabled?: boolean,
  max?: number,
  min?: number,
  placeholder?: string,
  readOnly?: boolean,
  step?: number,
};

export type FileControlOptions = {
  accept?: string,
  autofocus?: boolean,
  disabled?: boolean,
  multiple?: boolean,
  placeholder?: string,
  readOnly?: boolean,
  step?: number,
};

export type RangeControlOptions = {
  value: number,
  autofocus?: boolean,
  disabled?: boolean,
  labels?: boolean,
  marks?: boolean,
  max?: number,
  min?: number,
  placeholder?: string,
  readOnly?: boolean,
  step?: number,
};

export type ControlOptions = ButtonControlOptions
| TextControlOptions
| CheckboxControlOptions
| NumberControlOptions
| FileControlOptions
| TextAreaControlOptions
| RangeControlOptions;

export type ControlListenerUpdate = {
  id: string,
  key: string,
  value: string | boolean | number,
};

export type ControlListener = (update: ControlListenerUpdate) => void;
