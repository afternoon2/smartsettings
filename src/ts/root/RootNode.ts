import cuid from 'cuid';
import { PanelOptions } from '../panel/Panel';
import { ControlOptions, ControlListener } from '../controls/controls.types';

export abstract class RootNode {
  public id: string;
  public name: string;
  public listener: ControlListener | null;
  public options: PanelOptions | ControlOptions;

  constructor(
    name: string,
    options: PanelOptions | ControlOptions,
    listener: ControlListener | null,
    id: string | null,
  ) {
    const self = this;
    this.id = id || cuid();
    this.name = name;
    this.listener = listener;
    this.options = new Proxy(
      options,
      {
        set(
          // eslint-disable-next-line
          target: PanelOptions | ControlOptions,
          key: string,
          value: string | boolean | number,
        ) {
          if (self.listener) {
            self.listener({
              id: self.id,
              key,
              value,
            });
          }
          return true;
        }
      },
    ) as (PanelOptions | ControlOptions);
  }

  protected isValidElement(parent: HTMLElement | string | null): boolean {
    if (typeof parent === 'string') {
      return Boolean(document.querySelector(parent));
    }
    if (parent instanceof HTMLElement && document.body.contains(parent)) {
      return true;
    }
    return false;
  }

  protected getParentElement(parent: HTMLElement | string): HTMLElement {
    if (parent instanceof HTMLElement) {
      return parent;
    }
    return <HTMLElement>document.querySelector(parent);
  }
}