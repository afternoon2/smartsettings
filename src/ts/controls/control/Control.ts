import { RootNode, InternalState } from '../../root/RootNode';
import { ControlProps, ControlListener, ControlListenerUpdate } from './Control.types';

import Styles from '../../../sass/control.sass';
import Button from '../../../sass/button.sass';

export abstract class Control extends RootNode {
  public parentElement: HTMLElement;
  public element: HTMLElement;
  public abstract controlElement: HTMLElement;

  protected state: InternalState;
  protected listeners: Map<string, ControlListener> = new Map();

  private stateHandler: ProxyHandler<InternalState> = {
    set: this.createStateSetter(),
  }

  constructor(
    props: ControlProps,
    template: (state: InternalState) => string
  ) {
    super();
    this.state = this.createState(
      props.id, props.name, props.options,
      this.stateHandler,
    );
    this.parentElement = props.parentElement;
    this.element = this.createControlElement(template(this.state), this.state.name, this.state.id);
    this.parentElement.appendChild(this.element);
    this.listeners.set('invisible', this.onInvisible);
    this.listeners.set('disabled', this.onDisabled);
    this.listeners.set('readonly', this.onReadonlyChange);
    if (props.listener) {
      this.listeners.set('control', props.listener);
    }
    if (props.sectionListener) {
      this.listeners.set('section', props.sectionListener);
    }
    if (props.panelListener) {
      this.listeners.set('panel', props.panelListener);
    }
  }

  get readonly(): boolean {
    return this.state.readonly  as boolean;
  }

  set readonly(readonly: boolean) {
    this.state.readonly = readonly;
  }

  setListener(listener: ControlListener) {
    this.listeners.set('control', listener);
  }

  protected createRootDiv(): HTMLDivElement {
    const element = document.createElement('div');
    element.classList.add(Styles.control);
    return element;
  }

  protected createControlElement(template: string, name: string, id: string): HTMLDivElement {
    const element = this.createRootDiv();
    const main = document.createElement('main');
    const isButton = template.search(Button.button) > -1;
    main.classList.add(Styles.control__body);
    main.setAttribute('id', id);
    if (!isButton) {
      main.insertAdjacentHTML(
        'beforeend',
        `<p class="${Styles.control__name}" title="${name}">${name}</p>`,
      );
    }
    const content = document.createElement('div');
    content.classList.add(Styles.control__content);
    content.insertAdjacentHTML('beforeend', template);
    if (isButton) {
      content.classList.add(Styles['control__content--button']);
    }
    main.appendChild(content);
    element.appendChild(main);
    return element;
  }

  private onReadonlyChange: ControlListener = (update: ControlListenerUpdate) => {
    if (update.value === true) {
      if (!this.controlElement.hasAttribute('readonly')) {
        this.controlElement.setAttribute('readonly', 'true');
      }
    } else {
      if (this.controlElement.hasAttribute('readonly')) {
        this.controlElement.removeAttribute('readonly');
      }
    }
  }
}
