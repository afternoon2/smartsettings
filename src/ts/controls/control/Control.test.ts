import { Control } from './Control';
import { InternalState } from '../../root/RootNode';
import { ControlListener } from '../controls.types';

// @ts-ignore
import Styles from '../../../sass/control.sass';

class DerivedControl extends Control {
  public parentElement: HTMLElement = document.body;
  public element: HTMLElement;
  public controlElement: HTMLElement;

  protected state: InternalState;
  protected template = (state: InternalState): string => `
    <input type="text" id="testId" value="${state.value} />`;
  protected listeners: Map<string, ControlListener> = new Map();

  private stateHandler: ProxyHandler<InternalState> = {
    set: this.createStateSetter(),
  };

  constructor() {
    super();
    this.state = this.createState(
      'id', 'name', {}, this.stateHandler,
    );
    this.element = this.createControlElement(
      this.template(this.state),
      this.state.name,
      this.state.id,
    );
    this.controlElement = this.element.querySelector(`#testId`) as HTMLElement;
  }

  getRootDiv() {
    return this.createRootDiv();
  }

  getControlElement() {
    return this.createControlElement(
      this.template(this.state), this.state.name, this.state.id,
    );
  }
}

let instance: DerivedControl;

describe('Abstract Control class', () => {
  describe('Control class methods', () => {
    beforeEach(() => {
      instance = new DerivedControl();
    });
    
    test('createRootDiv', () => {
      const div = instance.getRootDiv();
      expect(div).toBeInstanceOf(HTMLDivElement);
      expect(div.classList.contains(Styles.control)).toBe(true);
    });

    test('createControlElement and withControlTemplate', () => {
      const controlElement = instance.getControlElement();
      expect(controlElement).toBeInstanceOf(HTMLElement);
      expect(controlElement).toMatchSnapshot();
    });
  });
});
