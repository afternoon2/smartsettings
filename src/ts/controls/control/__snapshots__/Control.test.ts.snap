// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Abstract Control class Control class methods createControlElement and withControlTemplate 1`] = `
<div
  class="control"
>
  
      
  <main
    class="control__body"
    id="id"
  >
    
        
    <p
      class="control__name"
      title="name"
    >
      
          name
        
    </p>
    
      
  </main>
  
      
  <div
    class="control__content"
  >
    
        
    
  </div>
</div>
`;
